src/PureSMT.hs

```diff
@@ -22,3 +29,59 @@ class Solve domain where
 
   -- |Solves a problem, producing an associated result with it
   solveProblem :: Problem domain result -> Solver -> IO result
+
+{-# NOINLINE solve #-}
+solve :: forall domain res . Solve domain => Ctx domain -> Problem domain res -> res
+solve ctx = unsafePerformIO $ do
+  -- Launch all our worker processes similar to how we did it in TypedProcess2.hs; but now
+  -- we end up with a list of MVars, which we will protect in another MVar.
+  allProcs <- launchAll @domain ctx >>= newMStack
+
+  -- Finally, we return the actual closure, the internals make sure
+  -- to use 'withMVar' to not mess up the command/response pairs.
+  return $ \problem -> unsafePerformIO $ do
+    ms <- popMStack allProcs
+    r <- withMVar ms $ \solver -> do
+      -- TODO: what happens in an exception? For now, we just loose a solver but we shouldn't
+      -- add it to the pool of workers and just retry the problem. In a future implementation
+      -- we could try launching it again
+      --pid <- unsafeSolverPid solver
+      --print pid
+      solveProblem @domain problem solver
+    pushMStack ms allProcs
+    return r
+
+launchAll :: forall domain . Solve domain => Ctx domain -> IO [MVar X.Solver]
+launchAll ctx = replicateM numCapabilities $ do
+  s <- X.launchSolverWithFinalizer "cvc4 --lang=smt2 --incremental --fmf-fun" debug0
+  initSolver @domain ctx s
+  newMVar s
+  where
+    -- TODO: these constants should become parameters at some point; the solver command too!
+
+    numCapabilities :: Int
+    numCapabilities = 3
```

Comment:
Why not start with the [RTS threading argument `numCapabilities`](https://hackage.haskell.org/package/base-4.14.1.0/docs/GHC-Conc.html#v:numCapabilities)?